### PR TITLE
TST: stats: add mpmath distribution

### PR DESCRIPTION
**Pull request by mdhaber**
Distribution class for creating tests and benchmarks.

### CODE REVIEW

File: scipy/stats/tests/mpmath_distribution.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
+import numpy as np
+from mpmath import mp
+
+mp.dps = 100  # default in case the user forgets to set it
+
+class Distribution:
```
Comment (mdhaber) [inline scipy/stats/tests/mpmath_distribution.py]: What should we call this? `ReferenceDistribution`?

File: scipy/stats/tests/mpmath_distribution.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+        def integrand(x):
+            return self._pdf(x, **kwargs)*(x - center)**order
+
+        if center is None:
+            center = self._mean(**kwargs)
+
+        a, b = self._support(**kwargs)
+        return mp.quad(integrand, (a, b))
+
+    def pdf(self, x):
```
Comment (mdhaber) [inline scipy/stats/tests/mpmath_distribution.py]: All of these are copy-paste boilerplate. There are a few ways to add them programmatically. Do you want me to implement one @tupui?

File: scipy/stats/tests/mpmath_distribution.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+        fun = np.vectorize(self._pdf)
+        x = self._make_mpf_array(x)
+        res = fun(x, **self._params)
+        return np.asarray(res, dtype=np.float64)[()]
```
Comment (mdhaber) [inline scipy/stats/tests/mpmath_distribution.py]: I think it would be useful to expose the output dtype as a parameter so that the full precision can be output if desired.

File: scipy/stats/tests/mpmath_distribution.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+    # always override __init__ so IDEs hint at shape parameter names
+    def __init__(self, *, mu, sigma):
+        super().__init__(mu=mu, sigma=sigma)
+
+    def _pdf(self, x, mu, sigma):
+        return mp.npdf(x, mu, sigma)
+
+class SkewNormal(Distribution):
+
+    def __init__(self, *, a):
```
Comment (mdhaber) [inline scipy/stats/tests/mpmath_distribution.py]: There's really no need to implement `loc` and `scale` parameters if we're using this to test the existing distributions. The reason is that `loc` and `scale` have generic implementations for all distributions, and that is already tested, so if we're interested in the accuracy of a distribution method, we only need to test it for varying *shape* parameters.

File: scipy/stats/tests/mpmath_distribution.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+    def _entropy(self, **kwargs):
+        def integrand(x):
+            logpdf = self._logpdf(x, **kwargs)
+            pdf = mp.exp(logpdf)
+            return  -pdf*logpdf
+
+        a, b = self._support(**kwargs)
+        return mp.quad(integrand, (a, b))
+
+    def _mean(self, **kwargs):
```
Comment (mdhaber) [inline scipy/stats/tests/mpmath_distribution.py]: I'm thinking about adding memoization to speed things up, but I'm going to keep it simple for now.

File: scipy/stats/tests/mpmath_distribution.py [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+        raise NotImplementedError("_pdf must be overridden.")
+
+    def _cdf(self, x, **kwargs):
+        a, _ = self._support(**kwargs)
+        return mp.quad(lambda x: self._pdf(x, **kwargs), (a, x))
+
+    def _sf(self, x, **kwargs):
+        _, b = self._support(**kwargs)
+        return mp.quad(lambda x: self._pdf(x, **kwargs), (x, b))
+
```
Comment (mdhaber) [inline scipy/stats/tests/mpmath_distribution.py]: `_ppf` and `_isf` will be added here.

File: scipy/stats/tests/test_mpmath_distribution.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    rng = np.random.default_rng(6716188855217730280)
+
+    x = rng.random(size=3)
+    a = rng.random(size=(2, 1))
+    rtol = 1e-15
+
+    dist = SkewNormal(a=a)
+    dist_ref = stats.skewnorm(a)
+
+    assert_allclose(dist.pdf(x), dist_ref.pdf(x), rtol=rtol)
```
Comment (mdhaber) [inline scipy/stats/tests/test_mpmath_distribution.py]: LMK if you want this parametrized @tupui.

**Review by mdhaber — COMMENTED** [review]
@tupui I thought I'd show you this draft. 

As you know, there has been a deluge of PRs improving the precision of distribution methods, and each one is accompanied by a little mpmath implementation of the distribution function to be tested.  Half of this code is boilerplate, e.g.
```python3
from mpmath import mp
mp.dps = 50
x = mp.mpf(...)
def fun(...):
```
The remainder that actually implements the method to be tested is usually unnecessary, too: because `mpmath` can be so accurate, we really only need to implement the PDF of each distribution, and reference values accurate to 16 digits can be generated by generic implementations of the methods most of the time.

Besides simplifying review and reducing bloat in the test files, I have some bigger things in mind. 
1. create accuracy benchmarks of distribution methods so we can see them improve over time
2. (distinct idea) characterize the accuracy of distribution methods. 

Re: 2, it would be very nice if we could make statements in the documentation like, 

> For arguments with magnitudes between 1e-10 and 1e10, this method is accurate to at least 10 digits of precision with 99.99% confidence.

I'd run Monte Carlo experiments to determine what statements we can make.

I had a few questions here before I open this up on the main repo with a demonstration of its use (i.e. using it to replace the reference values for some distributions in `test_distributions`). LMK:
- what the file structure should be (I doubt these should go straight in the `stats` test folder),
- what you think this should be called (https://github.com/mdhaber/scipy/pull/99#discussion_r1161499601),
- whether the public methods should be generated programmatically (https://github.com/mdhaber/scipy/pull/99#discussion_r1161499928), and 
- whether you want this test to be parametrized (https://github.com/mdhaber/scipy/pull/99#discussion_r1161502861). 
 
Other thoughts are also appreciated!

File: scipy/stats/tests/mpmath_distribution.py (hunk 2 above)
Comment (tupui) [reply scipy/stats/tests/mpmath_distribution.py]: I think it's ok. It's a bit of duplications but since it's testing code, I think it's better to keep it understandable, straightforward and without magic. It's then easier to see what is happening when we debug, etc. Also this way we are more confident with that and can go crazy if we want with other things.

File: scipy/stats/tests/mpmath_distribution.py (hunk 1 above)
Comment (tupui) [reply scipy/stats/tests/mpmath_distribution.py]: This, or `MPMathDistribution`.

File: scipy/stats/tests/mpmath_distribution.py [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+    # Minimalist distribution infrastructure for generating reference values
+    # of distribution functions. No edge case handling.
+
+    @staticmethod
+    def _make_mpf_array(x):
+        shape = np.shape(x)
+        x = np.asarray(x, dtype=np.float64).ravel()
+        return np.asarray([mp.mpf(xi) for xi in x]).reshape(shape)[()]
+
+    def __init__(self, **kwargs):
```
Comment (tupui) [inline scipy/stats/tests/mpmath_distribution.py]: Since this is for testing, is there some seeding/rng we could put there? Will matter if we add things like `random` (yes I refuse to call it `rvs` 😆).

File: scipy/stats/tests/test_mpmath_distribution.py (hunk 7 above)
Comment (tupui) [reply scipy/stats/tests/test_mpmath_distribution.py]: I don't think that would do much in terms of cleaning it, so fine like that I would say 👍

**Review by tupui — COMMENTED** [review]
That's a good idea! Will be better than what we do right now and allow more people to review with confidence 👍 (starting with me)

**Comment by tupui** [comment]
Forgot to say, I suppose this could be used in the work towards hypothesis testing? Did you have this in mind as well here?

File: scipy/stats/tests/mpmath_distribution.py (hunk 8 above)
Comment (mdhaber) [reply scipy/stats/tests/mpmath_distribution.py]: I don't think we will add `rvs`/`random`. If a distribution has a custom `rvs`, its output cannot be compared against the generic `rvs` we'd implement it. If a distribution relies on the default `rvs`, it does not need to be tested. So id on't see a need to add a `rng` at this time because it is all deterministic.
